### PR TITLE
Implement component-level testing framework

### PR DESCRIPTION
**Pull request by dariusgrassi**
The goal of this pull request is to implement a developer-friendly testing framework to verify components of the UPF are working. This framework uses [PTF](https://github.com/p4lang/PTF) to implement these test cases in Python.

We implement both unary and linerate testing options; much more context is given in `ptf/README` on these tests and much more, but essentially these options consist of either single-packet or high-speed tests.

For further context on how this framework is implemented, please refer to `ptf/README`.

### CODE REVIEW

**Comment by krsna1729** [comment]
Should the license be Apache-2.0 only in all of the new files?

**Comment by krsna1729** [comment]
there seems to be two png files that are not rendering properly. If possible have them as svg.

**Comment by krsna1729** [comment]
@pudelkoM can you test on your setups and review?

File: ptf/.env [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# Copyright 2021-present Open Networking Foundation
+# SPDX-License-Identifier: LicenseRef-ONF-Member-Only-1.0
```
Comment (pudelkoM) [inline ptf/.env:2]: Change license to apache 2. This is a leftover.

**Review by ccascone — APPROVED** [review]
@krsna1729 Max is on PTO, I just verified this together with @dariusgrassi. Tests are working and everything is set up for nightly runs on Jenkins. Here's a sample run (link to results is also available as a badge in the main README):
https://jenkins.onosproject.org/job/bess-upf-linerate-tests/6/

I'm going to merge this soon. Few improvements that we will tackle in later PRs:
- Simplify `ptf/config/docker_setup.sh` to only run steps necessary for PTF (e.g., remove `pfcpiface` since we are invoking the BESS gRPC API directly)
- Commit groovy pipeline to this repo (currently hosted in a different [repo](https://gerrit.onosproject.org/plugins/gitiles/ci-management/+/72a0d9b32bd08cbbbf181638d225bdfc47b69d06/jjb/pipeline/bess-upf-linerate.groovy))
- Execute ptf runner in the `BESS-testing` node instead of `menlo-pdp-lotta-nics`, to avoid generating the proto files twice

File: Makefile [hunk 2]
```diff
@@ -79,4 +79,10 @@ py-pb:
...
 	cp -a output/bess_pb/. ${PTF_PB_DIR}
 
-.PHONY: docker-build docker-push output pb
+fmt:
+	@gofmt -s -l -w $(GO_FILES)
+
+golint:
+	@docker run --rm -v $(CURDIR):/app -w /app/pfcpiface golangci/golangci-lint:latest golangci-lint run -v --config /app/.golangci.yml
+
+.PHONY: docker-build docker-push output pb fmt golint
```
Comment (ccascone) [inline Makefile:88]: @dariusgrassi should this change be here? We might have resolved merge conflicts in the wrong way

Comment (ccascone) [reply Makefile:88]: Nevermind... I was looking at one commit only, not the full PR diff. Please disregard.

**Comment by krsna1729** [comment]
ok

**Comment by ccascone** [comment]
`Pull Request / build (pull_request) ` is failing at the `Update PR with changes` step because of permission errors. I suspect this is happening because the PR originates from a fork. @krsna1729 can we ignore?

**Comment by krsna1729** [comment]
Yes those can be ignored

**Comment by krsna1729** [comment]
@dariusgrassi i should have looked closely - https://github.com/omec-project/upf-epc/pull/377/checks#step:9:8
Can you clean this up by shfmt?

https://github.com/omec-project/upf-epc/blob/6b1e5c18b7b4e2da62e002b8960ad802ba363dba/.github/workflows/pull_request.yml#L41-L46

**Pull request merged**